segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -1002,6 +1007,8 @@ private CompletableFuture<Void> adjustLastChunkLengths(MetadataTransaction txn)
                             ff = CompletableFuture.completedFuture(null);
                         }
                         return ff.thenApplyAsync(v -> {
+                            // Unset atomic writes as this version does not support atomic writes.
+                            segmentMetadata.setAtomicWrites(false);
```

Comment:
I don't understand this. `segmentMetadata` is constructed above with this set to true and has `checkInvariants()` invoked on it. But here the value is changed unconditionally and it is checked again. Can this be set earlier, perhaps when it is constructed in `get` on `MetadataTransaction`?

Reply:
System journals work by first reading a known snapshot of metadata about special 4 segments and then serially applying journal entries on top of that. `segment state= Snapshot + Journal records`

The 0.13 code will add a flag for atomic writes in the metadata to indicate whether to use reconcile logic or not.
So, snapshot written by 0.13 code will have that flag set in metadata.

This PR is relevant only when we are going from 0.13 back to 0.12 version.
This flag has no meaning because in 0.12 the writes are not atomic and expect to be reconciled.
Issue will happen when we go from 0.13 -> 0.12 and 0.13 again.
However, because we start with 0.13 snapshot, the metadata will still have that flag set but the code for 0.12 won't be writing atomically - expecting reconciliation to happen. When we eventually go to version 01.3 that version will see that flag set and skip reconciliation resulting in incorrect metadata/segment state.

Hence explicitly setting it false to avoid weird state transition errors.

Note that this PR is bare minimum code cherry-picked from 0.13 branch to deserialize the new "append" records and basically ignore them.

Reply:
In 0.12 between line 990 to 1004, the segmentMetadata will potentially change. 
Therefore, validating invariants at the end.